src/cuckoocache.h

```diff
@@ -0,0 +1,452 @@
+     */
+    uint32_t epoch_size;
+
+    /** hash_mask should be set to appropriately mask out a hash such that every
+     * masked hash is [0,size), eg, if floor(log2(size)) == 20, then hash_mask
+     * should be (1<<20)-1 
+     */
+    uint32_t hash_mask;
+
+    /** depth_limit determines how many elements insert should try to replace.
+     * Should be set to log2(n)*/
+    uint8_t depth_limit;
+
+
+    /** hash_function is a const instance of the hash function. It cannot be
+     * static or initialized at call time as it may have internal state (such as
+     * a nonce).
+     * */
+    const Hash hash_function;
+
+    /** compute_hashes is convenience for not having to write out this
+     * expression everywhere we use the hash values of an Element.
+     *
+     * @param e the element whose hashes will be returned
+     * @returns std::array<uint32_t, 8> of deterministic hashes derived from e
+     */
+    inline std::array<uint32_t, 8> compute_hashes(const Element& e) const
+    {
+        return {{hash_function.template operator()<0>(e) & hash_mask,
+                 hash_function.template operator()<1>(e) & hash_mask,
+                 hash_function.template operator()<2>(e) & hash_mask,
+                 hash_function.template operator()<3>(e) & hash_mask,
+                 hash_function.template operator()<4>(e) & hash_mask,
+                 hash_function.template operator()<5>(e) & hash_mask,
+                 hash_function.template operator()<6>(e) & hash_mask,
+                 hash_function.template operator()<7>(e) & hash_mask}};
+    }
+
+    /* end
+     * @returns a constexpr index that can never be inserted to */
+    constexpr uint32_t invalid() const
+    {
+        return ~(uint32_t)1;
+    }
+
+    /** allow_erase marks the element at index n as discardable. Threadsafe
+     * without any concurrent insert.
+     * @param n the index to allow erasure of
+     */
+    inline void allow_erase(uint32_t n) const
+    {
+        collection_flags.bit_set(n);
+    }
+
+    /** please_keep marks the element at index n as an entry that should be kept.
+     * Threadsafe without any concurrent insert.
+     * @param n the index to prioritize keeping
+     */
+    inline void please_keep(uint32_t n) const
+    {
+        collection_flags.bit_unset(n);
+    }
+
+    /** epoch_check handles the changing of epochs for elements stored in the
+     * cache. epoch_check should be run before every insert.
+     *
+     * First, epoch_check decrements and checks the cheap heuristic, and then does
+     * a more expensive scan if the cheap heuristic runs out. If the expensive
+     * scan suceeds, the epochs are aged and old elements are allow_erased. The
+     * cheap heuristic is reset to retrigger after the worst case growth of the
+     * current epoch's elements would exceed the epoch_size.
+     */
+    void epoch_check()
+    {
+        if ((--epoch_heuristic_counter) != 0)
```

Comment:
nit: technically this is an off-by-one: because you check this prior to actually doing the insert you'll always do an "expensive" scan twice for the first (and all) epochs. Practically, you might just want to set epoch_heuristic_count to something like epoch_size + 10 by default, though I suppose it doesnt matter all that much.